### PR TITLE
Persistence of options between script versions

### PR DESCRIPTION
**Pull request by Defimatt**
## Motivation and Context

This PR solves the problem of custom options getting overwritten when the script updates. Auto-updating the script from inside GreaseMonkey or TamperMonkey results in the options (which are stored directly in bot.user.js) getting overwritten with the defaults in the new script.

If you're like me, you like tweaking the options to your choice, and get annoyed when they get set back to default. Sure, you can copy and paste the options block from the previous script, but that requires you have it saved somewhere, which you probably don't when the script auto-updates.
## Description

How does it work?
- It uses LocalStorage to store a user's preferred options (via the persistence methods already built in to window.userInterface),
- window.bot.opt defined in the script now stores the **default** set of options, which are loaded only if no other options are available (i.e. this is the first time the user has used the bot at all, or the first time since this feature was included),
- The first time the script is run, the values in the script are saved,
- Every other time the script is run, the values are loaded, taking priority over the default options,
- If the user wants to change options, they make the changes in the new customBotOptions block at the start of the script. Any changes made here are picked up and saved next time the bot runs after saving the script,
- Here's the big difference: when the script is updated, it sees that there are saved options from a previous version, and they are loaded instead of the default options, thereby preserving the choices of the user from the previous version.

Here's the bad things:
- Users will have to learn to put any new settings they want in customBotOptions instead of window.bot.opt, otherwise they won't be saved between script versions. I've tried to make this obvious by putting customBotOptions at the top of the file, including a warning at the top of window.bot.opt and removing the explanatory comments from window.bot.opt,
- If the format of the option block changes, we have to change the code in two places.

I'm not saying this is the best way of doing this, just saying it's a start. If you have a better idea, please contribute! This is a stepping stone on the way to having a graphical options overlay (if we want that).

Also, this will have to be tested by a few people before it can be merged.
## Screenshots (if appropriate):

Flowchart
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: bot.user.js [hunk 1]
```diff
@@ -1308,6 +1332,30 @@ var userInterface = window.userInterface = (function () {
     userInterface.loadPreference('mobileRender', false);
     userInterface.loadPreference('leaderboard', true);
     window.nick.value = userInterface.loadPreference('savedNick', 'Slither.io-bot');
+    
+    // Load saved options, if any
+    var savedOptions = userInterface.loadPreference('options', null);    
+    if (null !== savedOptions) { // If there were saved options
```
Comment (Defimatt) [inline bot.user.js]: **REALLY!?** Meh, ok.

File: bot.user.js [hunk 2]
```diff
@@ -7,13 +7,47 @@ file, You can obtain one at http://mozilla.org/MPL/2.0/.
...
 // @match        http://slither.io/
 // @grant        none
 // ==/UserScript==
 
+/*
+Override bot options here
+Uncomment variables you wish to change from their default values
+Changes you make here will be kept between script versions
+*/
+var customBotOptions = {
```
Comment (Defimatt) [inline bot.user.js:22]: IDK what this means.

File: bot.user.js [hunk 3]
```diff
@@ -314,31 +348,21 @@ var bot = window.bot = (function () {
         sectorBox: {},
         currentFood: {},
         opt: {
-            // target fps
+            // These are the bot's default options
+            // If you wish to customise these, use
+            // customBotOptions above
```
Comment (ChadSki) [inline bot.user.js]: Why relocate this to another location? Customization is a good idea, but I don't like the idea of having two separate dicts with settings in them.

Why not just put the defaults into `customBotOptions` and uncomment some of those lines?

**Comment by ChadSki** [comment]
Hmm, I like the idea a lot but I'm not totally sold on having settings in two places...

File: bot.user.js (hunk 2 above)
Comment (ChadSki) [inline bot.user.js:22]: I think this is an error message due to Hound being configured slightly wrong.

@ErmiyaEskandary ? The keyword spacing rule looks for violations like `}else {`. This seems busted.

File: bot.user.js (hunk 1 above)
Comment (ChadSki) [inline bot.user.js]: Lol, linters can be aggressive at times

**Comment by ermiyaeskandary** [comment]
@ChadSki I'll have a look...

**Comment by ermiyaeskandary** [comment]
@ChadSki @MattDuffin  Should be fixed

**Comment by tjorim** [comment]
After some thought, I think this is indeed the best solution (for now).
LGTM :+1: (maybe we should add an option later to easily reset these options)

**Comment by Defimatt** [comment]
@ChadSki Trust me, I hated the inelegance of having code repeated (DRY principle violation) and I spent ages wrangling with alternative logic, but in the end my tiny brain couldn't think of anything else.

> Why relocate this to another location? Customization is a good idea, but I don't like the idea of having two separate dicts with settings in them.
> 
> Why not just put the defaults into customBotOptions and uncomment some of those lines?

If default options and custom options were stored in one place, when a new version of the script is installed, the logic would have no way of knowing that the options in the new script's defaults aren't the user's custom options.

I think there has to be an object with default settings (to be installed with the script should there not be any in localStorage already) and an object with custom settings (that take precedence over the defaults, and will be reloaded from localStorage even if the object with custom settings is removed from bot.user.js by a script update).

**Comment by Defimatt** [comment]
@tjorim Added

**Comment by ermiyaeskandary** [comment]
So this basically keeps your options even when the script updates? What are the negative aspects of this ?

**Comment by Defimatt** [comment]
@ErmiyaEskandary Correct. Here are the negative aspects I've identified:
- Users will have to learn to put any new settings they want in customBotOptions instead of window.bot.opt, otherwise they won't be saved between script versions. I've tried to make this obvious by putting customBotOptions at the top of the file, including a warning at the top of window.bot.opt and removing the explanatory comments from window.bot.opt,
- If the format of the option block changes, we have to change the code in two places.

**Comment by ermiyaeskandary** [comment]
@MattDuffin Ah...

> If the format of the option block changes, we have to change the code in two places.

This is easily outweighed by the positive aspects of this PR :+1: 

> Users will have to learn to put any new settings they want in customBotOptions instead of window.bot.opt, otherwise they won't be saved between script versions. I've tried to make this obvious by putting customBotOptions at the top of the file, including a warning at the top of window.bot.opt and removing the explanatory comments from window.bot.opt

So windows.bot.opt is the DEFAULT options and customBotOptions the USER DEFAULT options ? If so, it is then up to the user and we have already told them, despite slight confusion which may occur...

Approved! 

[Approved with PullApprove](https://pullapprove.com/ErmiyaEskandary/Slither.io-bot/pull-request/257/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ErmiyaEskandary/Slither.io-bot)

**Comment by Defimatt** [comment]
@ErmiyaEskandary have brought up to date with current master. I think I've done this correctly, but please confirm.

**Comment by ermiyaeskandary** [comment]
Correct but check the name ? Dev ?

**Comment by Defimatt** [comment]
@ErmiyaEskandary I wasn't sure what the version number would be, assuming 1.2.4.

Also, I think I broke the dog...

**Comment by ermiyaeskandary** [comment]
~~Please put the name as default and change the version to 1.2.4 ... I'll check the dog now ...~~

Thanks... I'll fix Hound now..

**Comment by ermiyaeskandary** [comment]
Should be fixed ... nvm

**Comment by tjorim** [comment]
Dog is broken because of 30eb21f0, because it thinks you want to change the .eslint-file (which isn't allowed). LGTM, we can just ignore the dog. :+1:

**Comment by ChadSki** [comment]
:+1: positives outweigh the negatives

[Approved with PullApprove](https://pullapprove.com/ErmiyaEskandary/Slither.io-bot/pull-request/257/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ErmiyaEskandary/Slither.io-bot)

**Comment by ChadSki** [comment]
@tjorim the sign-off string "LGTM" needs to be at the beginning of your comment for it to count.

**Comment by tjorim** [comment]
LGTM but I don't think it matters since I'm not on the approving list.

**Comment by ChadSki** [comment]
Well you ought to be, lol!

**Comment by ermiyaeskandary** [comment]
@tjorim Ah - that's why it didn't approve when you wrote LGTM... Adding you... How come you weren't on the list in the first place? :thought_balloon: lol

**Pull request merged**